nav2_shim_controller/include/nav2_shim_controller/shim_controller.hpp

```diff
@@ -0,0 +1,213 @@
+
+protected:
+  /**
+   * @brief Transforms global plan into same frame as pose, clips far away poses and possibly prunes passed poses
+   * @param pose pose to transform
+   * @return Path in new frame
+   */
+  nav_msgs::msg::Path transformGlobalPlan(const geometry_msgs::msg::PoseStamped & pose);
+
+  /**
+   * @brief Transform a pose to another frame.
+   * @param frame Frame ID to transform to
+   * @param in_pose Pose input to transform
+   * @param out_pose transformed output
+   * @return bool if successful
+   */
+  bool transformPose(
+    const std::string frame,
+    const geometry_msgs::msg::PoseStamped & in_pose,
+    geometry_msgs::msg::PoseStamped & out_pose) const;
+
+  /**
+   * @brief Whether robot should rotate to rough path heading
+   * @param carrot_pose current lookahead point
+   * @param angle_to_path Angle of robot output relatie to carrot marker
+   * @param angle_thresh Angle threshold
+   * @return Whether should rotate to path heading
+   */
+  bool shouldRotateToPath(
+    const geometry_msgs::msg::PoseStamped & carrot_pose, double & angle_to_path,
+    const double & angle_thresh);
+
+  /**
+   * @brief Whether robot should rotate to final goal orientation
+   * @param carrot_pose current lookahead point
+   * @return Whether should rotate to goal heading
+   */
+  bool shouldRotateToGoalHeading(const geometry_msgs::msg::PoseStamped & carrot_pose);
+
+  /**
+   * @brief Get lookahead point
+   * @param lookahead_dist Optimal lookahead distance
+   * @param path Current global path
+   * @return Lookahead point
+   */
+  geometry_msgs::msg::PoseStamped getLookAheadPoint(const double &, const nav_msgs::msg::Path &);
+
+  /**
+   * @brief Create a smooth and kinematically smoothed rotation command
+   * @param angular_vel angular velocity
+   * @param angle_to_path Angle of robot output relatie to carrot marker
+   * @param curr_speed the current robot speed
+   */
+  void rotateToHeading(
+    double & angular_vel,
+    const double & angle_to_path, const geometry_msgs::msg::TwistStamped & curr_speed);
+
+  /**
+   * @brief Сalculates the best command given the current pose and velocity and the global plan
+   *
+   * Depending on which plan is set and which angular threshold, this function decides whether the robot should
+   * turn to the planned path or the default plugin should be executed
+   *
+   * @param pose Current robot pose
+   * @param velocity Current robot velocity
+   * @return The best command for the robot to drive
+   */
+  geometry_msgs::msg::TwistStamped calcCmdVel(
+    const geometry_msgs::msg::PoseStamped & pose,
+    const geometry_msgs::msg::Twist & velocity);
+
+  std::shared_ptr<tf2_ros::Buffer> tf_;
+  std::string plugin_name_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros_;
+  nav2_costmap_2d::Costmap2D * costmap_;
+  rclcpp::Logger logger_ {rclcpp::get_logger("ShimController")};
+
+  double desired_linear_vel_;
```

Comment:
Many of these parameters should not exist, there's no need for this param, or look ahead dist, or most of them below. This controller only needs a param for max angular velocity / acceleration for rotating to heading / path.